migration: Add information about Vitess to TiDB migration (#5783)
pingcap/docs: migrate-from-vitess.md

Reviewer: Please add this doc to TOC.md.

Author: Done in a9350e2be78a7db5a82f0df14426332f4ebd8ffb

Reviewer: In this document, there are no detailed steps. Are these instructions sufficient enough for our users?

Author: I think this is ok. @mjonss what do you think?

Reviewer 2: Maybe we should add something like 'high level migration' to clarify it is not an step-by-step instruction?

# Vitess to TiDB migration

As the Vitess backend is based on MySQL you can use the regular MySQL migration tools like Dumpling, Lightning and DM which then need to be setup for each shard.

In addition to this there is a [Debezium connector for Vitess](https://debezium.io/documentation/reference/connectors/vitess.html) which allows you to use Kafka Connect or Apache Flink to stream changes from Vitess to TiDB.

As both Vitess and TiDB support the MySQL protocol and SQL dialect the amount of application level changes is expected to be small, except for things that directly manage sharding or other implementation specific things. To bridge the gap even more TiDB has support for the [`VITESS_HASH()`](/functions-and-operators/tidb-functions.md) function.

## Examples

### Dumpling and Lightning

![Vitess to TiDB Migration with TiDB backend](/media/vitess_to_tidb.png)

This is with Dumpling and Lightning where Lightning uses the `tidb` backend.

---

![Vitess to TiDB Migration with local backend](/media/vitess_to_tidb_dumpling_local.png)

This is with Dumpling and Lightning where Lightning uses the `local` backend to directly ingest data into TiKV.

### DM

> 

Reviewer: ```suggestion

The following example shows how DM migrates data from Vitess to TiDB.
```

---
title: Migrate Data from Vitess to TiDB
summary: Learn about the tools to migrate data from Vitess to TiDB. 
---

# Migrate Data from Vitess to TiDB

This document describes the tools that you can use to migrate data from [Vitess](https://vitess.io/) to TiDB. 

> As the backend of Vitess is based on MySQL, when migrating data from Vitess to TiDB, you can use the same migration tools that are applicable to MySQL, such as [Dumpling](/dumpling-overview.md), [TiDB Lightning](/tidb-lightning/tidb-lightning-overview.md), and [TiDB Data Migration (DM)](/dm/dm-overview.md).

Reviewer: ```suggestion
As the backend of Vitess is based on MySQL, when migrating data from Vitess to TiDB, you can use the same migration tools that are applicable to MySQL, such as [Dumpling](/dumpling-overview.md), [TiDB Lightning](/tidb-lightning/tidb-lightning-overview.md), and [TiDB Data Migration (DM)](/dm/dm-overview.md). Note that these tools should be set up for each shard in Vitess for data migration.
```

In addition to these tools, you can also use [Debezium connector for Vitess](https://debezium.io/documentation/reference/connectors/vitess.html). This connector enables you to use [Kafka Connect](https://kafka.apache.org/documentation/#connect) or [Apache Flink](https://nightlies.apache.org/flink/flink-docs-stable/) to stream changes from Vitess to TiDB.

Because both Vitess and TiDB support the MySQL protocol and SQL dialect, changes at the application level are expected to be small. For tasks directly managing sharding or other implementation-specific aspects, however, the changes might be larger. To facilitate the data migration from Vitess to TiDB, TiDB introduces the [`VITESS_HASH()`](/functions-and-operators/tidb-functions.md) function, which returns the hash of a string that is compatible with Vitess' HASH function.

## Examples

> ### Dumpling and Lightning

Reviewer: ```suggestion
### Dumpling and TiDB Lightning
```

The following two examples show how Dumpling and TiDB Lightning work together to migrate data from Vitess to TiDB.

> - In this example, TiDB Lightning uses the [logical import mode], which first encodes data into SQL statements and then runs the SQL statements to import data.

Reviewer: ```suggestion
- In this example, TiDB Lightning uses the [logical import mode](/tidb-lightning/tidb-lightning-logical-import-mode.md), which first encodes data into SQL statements and then runs the SQL statements to import data.
```

---
title: Migrate Data from Vitess to TiDB
summary: Learn about the tools to migrate data from Vitess to TiDB. 
---

# Migrate Data from Vitess to TiDB

This document describes the tools that you can use to migrate data from [Vitess](https://vitess.io/) to TiDB. 

> As the backend of Vitess is based on MySQL, when migrating data from Vitess to TiDB, you can use the same migration tools that are applicable to MySQL, such as [Dumpling](/dumpling-overview.md), [TiDB Lightning](/tidb-lightning/tidb-lightning-overview.md), and [TiDB Data Migration (DM)](/dm/dm-overview.md). Note that these tools should be set up for each shard in Vitess for data migration.

Reviewer 3: Here, it will be more appropriate to introduce to the users how to choose different solutions in different scenarios.
Since DM has already integrated Lightning physical mode since v7.6 (https://docs.pingcap.com/tidb/stable/task-configuration-file-full), we can encourage the user to use DM instead of Dumpling+Lightning+DM (for incremental data migration), it will be more easy and convenient. However, when the size of the data source is more than 10 TiB, it is recommended to use Dumpling + Lightning parallel physical mode+DM(incremental data migration) which is faster and more mature.

Reviewer 3: Rest LGTM

Author: Updated in a1127d2cd1d4524ed5745be1c3d59c6a0185aa96, PTAL

> As the backend of Vitess is based on MySQL, when migrating data from Vitess to TiDB, you can use the same migration tools that apply to MySQL, such as [Dumpling](/dumpling-overview.md), [TiDB Lightning](/tidb-lightning/tidb-lightning-overview.md), and [TiDB Data Migration (DM)](/dm/dm-overview.md). Note that these tools should be set up for each shard in Vitess for data migration. Generally, it is recommended to use DM with `task-mode` set to `all` and `import-mode` set to `physical`. If your data size exceeds 10 TiB, it is recommended to do the import in two steps, with Dumpling and TiDB Lightning as the first step to import existing data and DM as the second step to import incremental data.

Reviewer 4: ```suggestion
As the backend of Vitess is based on MySQL, when migrating data from Vitess to TiDB, you can use the same migration tools that apply to MySQL, such as [Dumpling](/dumpling-overview.md), [TiDB Lightning](/tidb-lightning/tidb-lightning-overview.md), and [TiDB Data Migration (DM)](/dm/dm-overview.md). Note that these tools should be set up for each shard in Vitess for data migration. 

Generally, before data migration, it is recommended to use DM with `task-mode` set to `all` and `import-mode` set to `physical`. For more information, see [Task configuration file template (advanced)](/dm/task-configuration-file-full.md#task-configuration-file-template-advanced). 

If your data size exceeds 10 TiB, it is recommended to do the import in two steps:

1. Use Dumpling and TiDB Lightning to import existing data.
2. Use DM to import incremental data.
```

Because both Vitess and TiDB support the MySQL protocol and SQL dialect, changes at the application level are expected to be small. For tasks directly managing sharding or other implementation-specific aspects, however, the changes might be larger. To facilitate the data migration from Vitess to TiDB, TiDB introduces the [`VITESS_HASH()`](/functions-and-operators/tidb-functions.md) function, which returns the hash of a string that is compatible with Vitess' HASH function.

## Examples

### Dumpling and TiDB Lightning

The following two examples show how Dumpling and TiDB Lightning work together to migrate data from Vitess to TiDB.

- In this example, TiDB Lightning uses the [logical import mode](/tidb-lightning/tidb-lightning-logical-import-mode.md), which first encodes data into SQL statements and then runs the SQL statements to import data.

    ![Vitess to TiDB Migration with TiDB backend](/media/vitess_to_tidb.png)

- In this example, TiDB Lightning uses the [physical import mode](/tidb-lightning/tidb-lightning-physical-import-mode.md) to directly ingest data into TiKV.

    ![Vitess to TiDB Migration with local backend](/media/vitess_to_tidb_dumpling_local.png)

### DM

> The following example shows how DM migrates data from Vitess to TiDB.

Reviewer 2: Maybe also have a link to DM here as well. Also the image should probably have arrows to both TiKV (the first data dump/reload is using similar physical mode as the second example for Lightning) and TiDB, where the binary log is applied through SQL as I understand?

This document describes the tools that you can use to migrate data from [Vitess](https://vitess.io/) to TiDB. 

> As the backend of Vitess is based on MySQL, when migrating data from Vitess to TiDB, you can use the same migration tools that are applicable to MySQL, such as [Dumpling](/dumpling-overview.md), [TiDB Lightning](/tidb-lightning/tidb-lightning-overview.md), and [TiDB Data Migration (DM)](/dm/dm-overview.md). Note that these tools should be set up for each shard in Vitess for data migration. The recommended method is to use DM with `task-mode` set to all and `import-mode` set to `physical`. When the data size is more than 10 TiB it is recommended to do the import in two steps, with Dumpling and TiDB Lightning being the first step and DM as the second step.

Reviewer: ```suggestion
As the backend of Vitess is based on MySQL, when migrating data from Vitess to TiDB, you can use the same migration tools that apply to MySQL, such as [Dumpling](/dumpling-overview.md), [TiDB Lightning](/tidb-lightning/tidb-lightning-overview.md), and [TiDB Data Migration (DM)](/dm/dm-overview.md). Note that these tools should be set up for each shard in Vitess for data migration. Generally, it is recommended to use DM with `task-mode` set to `all` and `import-mode` set to `physical`. If your data size exceeds 10 TiB, it is recommended to do the import in two steps, with Dumpling and TiDB Lightning as the first step to import existing data and DM as the second step to import incremental data.
```
